numalogic/udfs/trainer.py

```diff
@@ -0,0 +1,319 @@
+        if df.empty or not self._is_data_sufficient(payload, df):
+            _LOGGER.warning(
+                "%s - Insufficient data found for keys %s, shape: %s",
+                payload.uuid,
+                payload.composite_keys,
+                df.shape,
+            )
+            return Messages(Message.to_drop())
+        _LOGGER.info("%s - Data fetched, shape: %s", payload.uuid, df.shape)
+
+        # Construct feature array
+        x_train = self.get_feature_arr(df, payload.metrics)
+        _conf = self.get_conf(payload.config_id)
+
+        # Initialize artifacts
+        preproc_clf = self._construct_preproc_clf(_conf)
+        model = self._model_factory.get_instance(_conf.numalogic_conf.model)
+        thresh_clf = self._thresh_factory.get_instance(_conf.numalogic_conf.threshold)
+
+        # Train artifacts
+        artifacts = self.compute(
+            model,
+            x_train,
+            preproc_clf=preproc_clf,
+            threshold_clf=thresh_clf,
+            trainer_cfg=_conf.numalogic_conf.trainer,
+        )
+
+        # Save artifacts
+        # TODO perform multi-save here
+        self.save_artifact(
+            artifacts["preproc_clf"],
+            skeys=payload.composite_keys,
+            dkeys=[_conf.name for _conf in _conf.numalogic_conf.preprocess],
+            uuid=payload.uuid,
+        )
+        self.save_artifact(
+            artifacts["model"],
+            skeys=payload.composite_keys,
+            dkeys=[_conf.numalogic_conf.model.name],
+            uuid=payload.uuid,
+            train_size=x_train.shape[0],
+        )
+        self.save_artifact(
+            artifacts["threshold_clf"],
+            skeys=payload.composite_keys,
+            dkeys=[_conf.numalogic_conf.threshold.name],
+            uuid=payload.uuid,
+        )
+
+        _LOGGER.debug(
+            "%s - Time taken in trainer: %.4f sec", payload.uuid, time.perf_counter() - _start_time
+        )
+        return Messages(Message.to_drop())
+
+    @staticmethod
+    def _construct_preproc_clf(_conf: StreamConf) -> Optional[artifact_t]:
+        preproc_factory = PreprocessFactory()
+        preproc_clfs = []
+        for _cfg in _conf.numalogic_conf.preprocess:
+            _clf = preproc_factory.get_instance(_cfg)
+            preproc_clfs.append(_clf)
+        if not preproc_clfs:
+            return None
+        if len(preproc_clfs) == 1:
+            return preproc_clfs[0]
+        return make_pipeline(*preproc_clfs)
+
+    def save_artifact(
+        self, artifact: artifact_t, skeys: list[str], dkeys: list[str], uuid: str, **metadata
+    ) -> None:
+        """
+        Save artifact to the registry.
+
+        Args:
+            artifact: Artifact to save
+            skeys: List of keys
+            dkeys: List of dkeys
+            uuid: UUID
+            **metadata: Additional metadata
+        """
+        if not artifact:
+            return
+        if isinstance(artifact, StatelessTransformer):
```

Comment:
For preproc maybe we still might to add checks cause it will return a sklearn pipeline object?